tap_googleads/utils.py

```diff
@@ -0,0 +1,34 @@
+"""Utility functions for tap functionality"""
+
+from typing import Optional
+
+
+def replicate_pk_at_root(data: dict, jsonpaths: Optional[list]) -> dict:
+    """Replicates the indicated values as root properties.
+
+    Arguments:
+        data: The json object to be modified.
+        jsonpaths: A list of json paths in format 'level1.level2.etc' that
+            indicate which values are to be replicated as a root level property.
+
+    Returns:
+        A modified json record with the duplicate values as root level primary keys.
+    """
+    if not jsonpaths:
+        return data
+
+    pk_str = ""
+    new_data = data.copy()
+    for path in jsonpaths:
+        levels = path.split(".")
+
+        # recursively build pk value
+        val = data.copy()
+        for i, level in enumerate(levels):
+            val = val[level]
+            if i == len(levels) - 1:
+                pk_str += ":" + str(val) if pk_str else str(val)
```

Comment:
I know for mssql there's a limit to the size a PK index can be. https://docs.microsoft.com/en-us/sql/relational-databases/tables/primary-and-foreign-key-constraints?view=sql-server-ver15#:~:text=A%20table%20can%20contain%20only,key%20length%20of%20900%20bytes. 

Postgres has some similar length I believe. I think we're fine for now, but we may have to evaluate this at some point and maybe just md5 the whole thing